verilog/analysis/checkers/struct_union_name_style_rule.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright 2017-2019 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/struct_union_name_style_rule.h"
+
+#include <string>
+#include <vector>
+
+#include "absl/strings/str_cat.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/bound_symbol_manager.h"
+#include "common/strings/naming_utils.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/CST/type.h"
+#include "verilog/analysis/lint_rule_registry.h"
+
+namespace verilog {
+namespace analysis {
+
+VERILOG_REGISTER_LINT_RULE(StructUnionNameStyleRule);
+
+using verible::GetStyleGuideCitation;
+using verible::LintRuleStatus;
+using verible::LintViolation;
+using verible::SyntaxTreeContext;
+
+absl::string_view StructUnionNameStyleRule::Name() {
+  return "struct-union-name-style";
+}
+const char StructUnionNameStyleRule::kTopic[] = "struct-union-conventions";
+const char StructUnionNameStyleRule::kMessageStruct[] =
+    "Struct names must use lower_snake_case naming convention and end with _t.";
+const char StructUnionNameStyleRule::kMessageUnion[] =
+    "Union names must use lower_snake_case naming convention and end with _t.";
+
+std::string StructUnionNameStyleRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat(
+      "Checks that ",
+      Codify("struct", description_type),
+      " and ",
+      Codify("union", description_type),
+      " names use lower_snake_case naming convention and end with '_t'. See ",
+      GetStyleGuideCitation(kTopic), ".");
+}
+
+void StructUnionNameStyleRule::HandleSymbol(const verible::Symbol& symbol,
+                                       const SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  if (matcher_typedef_.Matches(symbol, &manager)) {
+    const char* msg;
+    if (!FindAllStructDataTypeDeclarations(symbol).empty()) {
```

Comment:
Alternative to searching inside a type declaration for the underlying struct/union type, if you shape nodes of `kTypeDeclaration` to have consistent structure (children positions), you could directly reference a child by index using:
https://github.com/google/verible/blob/d34420461f422d1cdbf4585a376b5a3210b9f828/common/text/tree_utils.h#L69

Reply:
It could be done, but I think a helper function would be needed anyway. It would certainly be faster, though.

Reply:
Leave a TODO then.

Reply:
Done.